### PR TITLE
nixos/test-driver: additional Python packages in test driver

### LINKED ISSUE #172325: nixosTest: pass additional python libraries to testScript
## Issue description

I would to like to pass additional `python` libraries to the version of `python` which executes the `testScript` script in a nixos test.

cc @tfc

### PR DESCRIPTION
**Pull request by tljuniper**
###### Description of changes

This adds an additional optional parameter to the test driver that allows for adding custom Python packages to be used by the test driver code.

There's also a new section in the docs that explains how to use this.

Fixes #172325.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tljuniper** [comment]
@ofborg test additional-python-packages login

File: nixos/doc/manual/development/writing-nixos-tests.section.md [hunk 1]
```diff
@@ -380,3 +380,24 @@ with foo_running:
...
+
+## Adding Python packages to the test script {#ssec-python-packages-in-test-script}
+
+When additional Python libraries are required in the test script, they can be
+added using the parameter `withAdditionalPythonPackages`. For example, you could add
+`numpy` like this:
+
+```py
+import ./make-test-python.nix {
+  withAdditionalPythonPackages = p: [p.numpy];
```
Comment (roberth) [inline nixos/doc/manual/development/writing-nixos-tests.section.md:392]: I haven't seen `with*` in as a parameter before. How about `extraPythonPackages`?

File: nixos/lib/testing-python.nix [hunk 2]
```diff
@@ -52,12 +52,13 @@ rec {
     , skipLint ? false
     , passthru ? {}
     , interactive ? false
+    , withAdditionalPythonPackages ? (_ :[])
```
Comment (roberth) [inline nixos/lib/testing-python.nix:55]: I plan to simplify such wiring with the module system. Let's get your change in first.

File: nixos/tests/additional-python-packages.nix [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import ./make-test-python.nix ({ ... }:
+{
+  name = "additional-python-packages";
+
+  withAdditionalPythonPackages = p: [p.numpy];
+
+  nodes.machine = {...}: {};
```
Comment (roberth) [inline nixos/tests/additional-python-packages.nix:7]: Who needs VMs, right?

I guess `nodes = { };` didn't work?

File: nixos/doc/manual/development/writing-nixos-tests.section.md (hunk 1 above)
Comment (tljuniper) [reply nixos/doc/manual/development/writing-nixos-tests.section.md]: `extra` sounds much nicer, thanks for the suggestion. I've renamed the tests as well to match the new name and updated the docs.

File: nixos/tests/additional-python-packages.nix (hunk 3 above)
Comment (tljuniper) [reply nixos/tests/additional-python-packages.nix]: No VMs --> probably the fastest nixos test yet! :grin: 

TBH, I didn't try that. It does work---I've fixed it here and in the docs.

**Comment by tljuniper** [comment]
@ofborg test additional-python-packages login

**Comment by aanderse** [comment]
LGTM... How about you @roberth?

**Review by roberth — COMMENTED** [review]
No objections.

One question though; is `numpy` a realistic example here? I'd expect people to import testing-related libraries, or perhaps some application-specific thing. Certainly not a blocker for the test. If we can improve the docs, that would be nice.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Still curious why numpy

**Comment by aanderse** [comment]
Thanks everyone!

**Comment by tljuniper** [comment]
Thanks everyone as well!

> Still curious why numpy

It's rather well known and allows for a nice one-liner to check if the lib is there. We could change it so something more test-related though. I have a small typo-fix PR coming up for some docs, that might be a good place to squeeze in that change.

Sorry for not replying earlier; I didn't have access to my desk the past few days.

**Pull request merged**